### DISCUSSION TITLE
Multiple running timers

### DISCUSSION THREAD

**Discussion by dallyger** [solidtime-io/solidtime#302] [Feature requests]
Sometimes it is necessary to have multiple timers running simultaneously. Currently starting another timer requires the first one to be stopped.

Is this something you may implement in the future?

**Comment by Onatcer** [comment]
what is your use case for two simultaneous timers? we plan to support tracking of attendance/work time separately from the time spent on projects in the future, so those would be able to run at the same time.

**Reply by dallyger** [reply]
This sounds interesting. And I and sorry. I should have included a use case directly.

An example would be working for an MSP managing the IT infrastructure for multiple clients.
Sometimes having to do some generic maintenance on multiple clients.
All of those tasks having low mental overhead and are doable remotely and simultaneously.

Examples of those tasks, which could also be mixed at the same time:
- installing OS updates
- installing software
- checking backups were successful
- taking support calls (and they have to gather additional information)
- ... and probably more I cannot think of right now.

So remoteing into client A updating multiple servers is mostly waiting for download/install to finish and clicking next/confirm a couple of times. While waiting, you can do the same for clients B, C, and D. Each time there is no immediate input required, switching to the next client.

Due to it affecting multiple clients, they have to be invoiced separately.
The waiting itself is charged too, because it blocks any more demanding tasks in that time.
However, the tasks themselves may not start at the same time and will most likely differ in their time to completion. So a single generic timer is not an option.

Therefore, two timers are probably not enough either. More like a dynamic stack of timers, which can be started and stopped individually.

**Reply by KiljanK** [reply]
Not the original commenter, but i'd also love it if this was a possibility. Projects overlap sometimes, and having the flexibility to run multiple timers simultaneously would be a godsend.

Sometimes it's nice to track _"usage time of specific server / device / lab equipment / tool"_ and also track the project which said thing was used for.

Side note: i don't know if this is a valid assessment, but i feel like it's best to generally think of this feature-request as "multi-timer", instead of just "two timers". The general possibility of having any amount of timers (within reason) seems great for product flexibility, if implementing an attendance timer is already on the roadmap anyways. Seems useful to build the "base" multi-timer feature and have the attendance timer be built on top of that.

Also: i believe i've seen some clockify forum posts asking for multiple timers (to little avail), so it might also be a selling point for those types of users, or at lesst a selling point for using solidtime instead of clockify in general.

**Reply by Onatcer** [reply]
my concerns about implementing multiple timers are primarily about UX i think it will get really confusing that's why i try to understand what you would use multiple timers for. If you want to f.e. track the time some tool / device / equipment is used I think you could tags for that. 

Having multiple timers as a feature will likely also result in confusing metrics. F.e. if you have 4 time entries running at the same time for 2 hours, your statistics everywhere would show that you worked for 8 hours that day. 

Attandance timers would be different because they are here to track working time in a much simpler way as Clock-in/Clock-out. Their primary use is to keep track of all time spent within an employment contract not just time spent on specific projects or clients.

**Reply by dallyger** [reply]
> Having multiple timers as a feature will likely also result in confusing metrics. F.e. if you have 4 time entries running at the same time for 2 hours, your statistics everywhere would show that you worked for 8 hours that day.

This is the expected behavior. There are days with 20+ billable hours in a normal work day.
So multiple timers should have nothing to do with an attendance timer.
Maybe tracking start of first timer to end of last timer. Or maybe, it is simply another one of the multi timers but with an "attendance" tag or something.

**Reply by KiljanK** [reply]
> my concerns about implementing multiple timers are primarily about UX i think it will get really confusing that's why i try to understand what you would use multiple timers for.

Oh i wasn't suggesting that it should be the default behaviour. I think it would be cool if it was _possible_. Ideally, a user would get the prompt _"do you want this timer to run simultaneously, or do you want to end the previous timer first?"_ the very first time they try to start a timer while another one is already running.

Offering the possibility to enable simultaneous timers with the option to change the setting and max amount of concurrent timers later on could solve the issue of UX imo.

> If you want to f.e. track the time some tool / device / equipment is used I think you could tags for that.

I'm not necessarily using the tool from start to finish of the currently running activity, so that wouldn't be possible, sadly. And entering the time manually just circles back to the base idea of why timers are nicer to use than manual entries.

> Having multiple timers as a feature will likely also result in confusing metrics. F.e. if you have 4 time entries running at the same time for 2 hours, your statistics everywhere would show that you worked for 8 hours that day. 

Honestly, that's not an issue for me, since i intend to use solidtime not only for work, but also for for private activities where i do not need to bill the hours.

And i think it's probably quite easy to a have a checkbox that toggles between "total time" and "cumulative time" during calculatuon, but i might be wrong, i haven't had a chance to look at the repo that much yet.

As @dallyger had noted already, it might also be the behaviour users want. If someone is billing a client for `project time + energy cost of a specific tool` for instance, it'd be great to track both simultaneously at different rates.

# However,

I understand that this might be too niche, and i do not want to slow down the development of this project with requests to features you do not deem useful. I really really really like what you're doing, especially the fact that it's open-source, so i'm already happy that you're doing this at all.

That being said, do you think it'd be possible for me to implement multi-timers myself in case i get around to tinkering with the codebase someday, or is the singular-timer part too deeply ingrained in the codebase?
